src/recon_buildblock/PLSPrior.cxx

```diff
@@ -0,0 +1,607 @@
+                                          const DiscretisedDensity<3,elemT> &image_grad_z,
+                                          const DiscretisedDensity<3,elemT> &image_grad_y,
+                                          const DiscretisedDensity<3,elemT> &image_grad_x){
+
+    const int min_z = image_grad_x.get_min_index();
+    const int max_z = image_grad_x.get_max_index();
+
+
+        for (int z=min_z; z<=max_z; z++)
+          {
+
+            const int min_y = image_grad_x[z].get_min_index();
+            const int max_y = image_grad_x[z].get_max_index();
+
+
+
+              for (int y=min_y;y<= max_y;y++)
+                {
+
+                  const int min_x = image_grad_x[z][y].get_min_index();
+                  const int max_x = image_grad_x[z][y].get_max_index();
+
+
+
+                    for (int x=min_x;x<= max_x;x++)
+                    {
+                     norm_im_grad[z][y][x]   = sqrt (square(image_grad_z[z][y][x]) + square(image_grad_y[z][y][x]) + square(image_grad_x[z][y][x]) + square(this->eta));
+                    }
+                 }
+          }
+}
+
+template <typename elemT>
+void
+PLSPrior<elemT>::
+compute_inner_product_and_penalty(DiscretisedDensity<3,elemT> &inner_product,
+                                  DiscretisedDensity<3,elemT> &penalty,
+                                  DiscretisedDensity<3,elemT> &pet_im_grad_z,
+                                  DiscretisedDensity<3,elemT> &pet_im_grad_y,
+                                  DiscretisedDensity<3,elemT> &pet_im_grad_x,
+                      const DiscretisedDensity<3,elemT> &pet_image){
+
+
+
+
+    compute_image_gradient_element (pet_im_grad_z,0,pet_image);
+    compute_image_gradient_element (pet_im_grad_y,1,pet_image);
+    compute_image_gradient_element (pet_im_grad_x,2,pet_image);
+
+
+    const int min_z = pet_image.get_min_index();
+    const int max_z = pet_image.get_max_index();
+
+
+        for (int z=min_z; z<=max_z; z++)
+          {
+
+            const int min_y = pet_image[z].get_min_index();
+            const int max_y = pet_image[z].get_max_index();
+
+
+
+              for (int y=min_y;y<= max_y;y++)
+                {
+
+                  const int min_x = pet_image[z][y].get_min_index();
+                  const int max_x = pet_image[z][y].get_max_index();
+
+
+
+                    for (int x=min_x;x<= max_x;x++)
+                    {
+                     inner_product[z][y][x]   = ((pet_im_grad_z[z][y][x]*(*anatomical_grad_z_sptr)[z][y][x]/(*get_norm_sptr())[z][y][x]) +
+                                                 (pet_im_grad_y[z][y][x]*(*anatomical_grad_y_sptr)[z][y][x]/(*get_norm_sptr())[z][y][x]) +
+                                                 (pet_im_grad_x[z][y][x]*(*anatomical_grad_x_sptr)[z][y][x]/(*get_norm_sptr())[z][y][x]));
+
+                     penalty[z][y][x]= sqrt (this->alpha*this->alpha + pet_im_grad_z[z][y][x]*pet_im_grad_z[z][y][x] +
```

Comment:
can use `square(pet_im_grad_z[z][y][x])`. might be a bit faster (and looks simpler).
(`square` is a STIR inline)